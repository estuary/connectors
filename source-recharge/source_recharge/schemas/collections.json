{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "required": ["id"],
  "properties": {
    "id": {
      "type": ["integer"]
    },
    "name": {
      "type": ["null", "string"]
    },
    "created_at": {
      "type": ["null", "string"],
      "format": "date-time"
    },
    "updated_at": {
      "type": ["null", "string"],
      "format": "date-time"
    }
  }
}
