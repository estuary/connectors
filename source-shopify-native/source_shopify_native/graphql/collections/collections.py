from datetime import datetime
from logging import Logger
import json
from typing import Any, AsyncGenerator, ClassVar, Literal

from ...models import ShopifyGraphQLResource, SortKey


class _Collections(ShopifyGraphQLResource):
    COLLECTION_TYPE: ClassVar[str] = ""
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT
    QUERY = """
    title
    handle
    description
    descriptionHtml
    templateSuffix
    sortOrder
    productsCount {
        count
        precision
    }
    products {
        edges {
            node {
                id
                legacyResourceId
            }
        }
    }
    publications {
        edges {
            node {
                __typename
                publishDate
                publication {
                    id
                    name
                }
            }
        }
    }
    """

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        if not cls.COLLECTION_TYPE:
            raise NotImplementedError("Subclasses must set COLLECTION_TYPE")

    @staticmethod
    def build_query_with_type(
        collection_type: Literal["custom", "smart"], start: datetime, end: datetime
    ) -> str:
        return _Collections.build_query_with_fragment(
            start,
            end,
            query=f"collection_type:{collection_type}",
            includeCreatedAt=False,
        )

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        raise NotImplementedError("build_query method must be implemented")

    @staticmethod
    async def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        PRODUCTS_KEY = "products"
        PUBLICATIONS_KEY = "publications"
        current_collection = None

        async for line in lines:
            record: dict[str, Any] = json.loads(line)
            id: str = record.get("id", "")
            parent_id = record.get("__parentId", "")
            typename = record.get("__typename", "")

            if "gid://shopify/Collection/" in id:
                if current_collection:
                    yield current_collection

                current_collection = record
                current_collection[PRODUCTS_KEY] = []
                current_collection[PUBLICATIONS_KEY] = []

            elif "gid://shopify/Product/" in id:
                if not current_collection:
                    log.error("Found a product before finding a collection.")
                    raise RuntimeError()
                if parent_id != current_collection.get("id", ""):
                    log.error(
                        "Product's parent id does not match the current collection's id.",
                        {
                            "product.id": id,
                            "product.__parentId": parent_id,
                            "current_collection.id": current_collection.get("id"),
                        },
                    )
                    raise RuntimeError()

                current_collection[PRODUCTS_KEY].append(record)

            elif typename == "CollectionPublication":
                if not current_collection:
                    log.error("Found a publication before finding a collection.")
                    raise RuntimeError()
                if parent_id != current_collection.get("id", ""):
                    log.error(
                        "Publication's parent id does not match the current collection's id.",
                        {
                            "publication.__parentId": parent_id,
                            "current_collection.id": current_collection.get("id"),
                        },
                    )
                    raise RuntimeError()

                record.pop("__typename", None)
                current_collection[PUBLICATIONS_KEY].append(record)

            else:
                log.error("Unidentified line in JSONL response.", record)
                raise RuntimeError()

        if current_collection:
            yield current_collection


class CustomCollections(_Collections):
    NAME = "custom_collections"
    COLLECTION_TYPE = "custom"

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        return _Collections.build_query_with_type("custom", start, end)

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return _Collections.process_result(log, lines)


class SmartCollections(_Collections):
    NAME = "smart_collections"
    COLLECTION_TYPE = "smart"

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        return _Collections.build_query_with_type("smart", start, end)

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return _Collections.process_result(log, lines)
