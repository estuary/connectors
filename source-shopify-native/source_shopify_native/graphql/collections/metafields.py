from datetime import datetime
from logging import Logger
from typing import AsyncGenerator
from ..metafields import MetafieldsResource

from source_shopify_native.models import SortKey


class CustomCollectionMetafields(MetafieldsResource):
    NAME = "custom_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return CustomCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:custom",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, CustomCollectionMetafields.PARENT_ID_KEY
        )


class SmartCollectionMetafields(MetafieldsResource):
    NAME = "smart_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return SmartCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:smart",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, SmartCollectionMetafields.PARENT_ID_KEY
        )
