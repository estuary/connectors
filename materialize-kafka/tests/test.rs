use core::str;
use std::collections::HashMap;

use rdkafka::{
    admin::{AdminClient, AdminOptions},
    consumer::{BaseConsumer, Consumer},
    ClientConfig, Message, TopicPartitionList,
};

const BOOTSTRAP_SERVERS: &str = "localhost:9092";
const SCHEMA_REGISTRY_ENDPOINT: &str = "http://localhost:8081";
static KAFKA_INTERNAL_TOPICS: [&str; 4] = [
    "__consumer_offsets",
    "__amazon_msk_canary",
    "_schemas",
    "__transaction_state",
];

#[test]
fn test_spec() {
    let output = std::process::Command::new("flowctl")
        .args([
            "raw",
            "spec",
            "--source",
            "tests/test.flow.yaml",
            "--name",
            "acmeCo/materialize-kafka/avro",
        ])
        .output()
        .unwrap();

    assert!(output.status.success());
    let got: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap();
    insta::assert_snapshot!(serde_json::to_string_pretty(&got).unwrap());
}

#[tokio::test]
async fn test_materialization() {
    drop_topics().await;

    for name in [
        "acmeCo/materialize-kafka/avro",
        "acmeCo/materialize-kafka/json",
    ] {
        let output = std::process::Command::new("flowctl")
            .args([
                "preview",
                "--source",
                "tests/test.flow.yaml",
                "--fixture",
                "tests/fixture.json",
                "--name",
                name,
            ])
            .output()
            .unwrap();

        println!("{}", str::from_utf8(&output.stderr).unwrap());
        assert!(output.status.success());
    }

    insta::assert_snapshot!(snapshot_topics().await);
}

async fn drop_topics() {
    let admin: AdminClient<_> = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .create()
        .unwrap();

    let topics = list_topics().await;
    let topics = topics.iter().map(String::as_str).collect::<Vec<&str>>();
    if !topics.is_empty() {
        admin
            .delete_topics(&topics, &AdminOptions::default())
            .await
            .unwrap();
    }
}

async fn list_topics() -> Vec<String> {
    let consumer: BaseConsumer = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .create()
        .unwrap();

    let meta = consumer.fetch_metadata(None, None).unwrap();
    let mut topics = meta
        .topics()
        .iter()
        .filter(|t| !KAFKA_INTERNAL_TOPICS.contains(&t.name()))
        .map(|t| t.name().to_string())
        .collect::<Vec<_>>();
    topics.sort();
    topics
}

async fn snapshot_topics() -> String {
    let mut out = String::new();

    for topic in list_topics().await {
        if !out.is_empty() {
            out.push_str("\n");
        }
        out.push_str("************\n");
        out.push_str(format!("Topic: {}\n", topic).as_str());
        out.push_str("************\n");
        out.push_str(&snapshot_topic(&topic).await);
        out.push_str("\n");
    }

    out
}

async fn snapshot_topic(topic: &str) -> String {
    let consumer: BaseConsumer = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .set("enable.auto.commit", "false")
        .set("group.id", "materialize-kafka")
        .set("auto.offset.reset", "earliest")
        .set("enable.partition.eof", "true")
        .create()
        .unwrap();

    let meta = consumer.fetch_metadata(Some(topic), None).unwrap();
    assert!(meta.topics()[0].partitions().len() != 0);
    let assignment =
        meta.topics()[0]
            .partitions()
            .iter()
            .fold(TopicPartitionList::new(), |mut tpl, p| {
                tpl.add_partition(topic, p.id());
                tpl
            });
    consumer.assign(&assignment).unwrap();
    let mut n = consumer.assignment().unwrap().count();

    let mut schema_cache: HashMap<u32, apache_avro::Schema> = HashMap::new();

    let mut rows = Vec::new();
    for msg in consumer.iter() {
        match msg {
            Ok(msg) => {
                let key = parse_datum(msg.key().unwrap(), &mut schema_cache).await;
                let value = parse_datum(msg.payload().unwrap(), &mut schema_cache).await;
                let row = format!("key: {}, value: {}", key, value);
                rows.push(row);
            }
            _ => {
                n -= 1;
                if n == 0 {
                    break;
                }
            }
        }
    }
    rows.sort();
    rows.join("\n")
}

async fn parse_datum(datum: &[u8], schema_cache: &mut HashMap<u32, apache_avro::Schema>) -> String {
    match datum[0] {
        0 => {
            let schema_id = u32::from_be_bytes(datum[1..5].try_into().unwrap());

            if !schema_cache.contains_key(&schema_id) {
                schema_cache.insert(schema_id, fetch_schema(schema_id).await);
            }

            let schema = schema_cache.get(&schema_id).unwrap();
            let avro_value = apache_avro::from_avro_datum(schema, &mut &datum[5..], None).unwrap();

            source_kafka::pull::avro_to_json(avro_value, schema)
                .unwrap()
                .to_string()
        }
        _ => str::from_utf8(datum).unwrap().to_owned(),
    }
}

async fn fetch_schema(id: u32) -> apache_avro::Schema {
    let schema_string = reqwest::get(format!(
        "{SCHEMA_REGISTRY_ENDPOINT}/schemas/ids/{id}/schema"
    ))
    .await
    .unwrap()
    .error_for_status()
    .unwrap()
    .text()
    .await
    .unwrap();

    apache_avro::Schema::parse_str(&schema_string).unwrap()
}
