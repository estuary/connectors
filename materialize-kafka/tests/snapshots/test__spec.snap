---
source: tests/test.rs
expression: "serde_json::to_string_pretty(&got).unwrap()"
---
{
  "configSchema": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {
      "bootstrap_servers": {
        "description": "The initial servers in the Kafka cluster to initially connect to, separated by commas. The Kafka client will be informed of the rest of the cluster nodes by connecting to one of these nodes.",
        "order": 0,
        "title": "Bootstrap Servers",
        "type": "string"
      },
      "compression": {
        "default": "lz4",
        "description": "Compression algorithm to use for messages. Note that not all Kafka brokers support all compression algorithms.",
        "enum": [
          "none",
          "gzip",
          "lz4",
          "snappy",
          "zstd"
        ],
        "order": 4,
        "title": "Compression",
        "type": "string"
      },
      "credentials": {
        "description": "The connection details for authenticating a client connection to Kafka via SASL. When not provided, the client connection will attempt to use PLAINTEXT (insecure) protocol. This must only be used in dev/test environments.",
        "discriminator": {
          "propertyName": "auth_type"
        },
        "oneOf": [
          {
            "properties": {
              "auth_type": {
                "const": "user_password",
                "default": "user_password",
                "order": 0,
                "type": "string"
              },
              "mechanism": {
                "default": "PLAIN",
                "description": "The SASL Mechanism describes how to exchange and authenticate clients/servers.",
                "enum": [
                  "PLAIN",
                  "SCRAM-SHA-256",
                  "SCRAM-SHA-512"
                ],
                "order": 1,
                "title": "SASL Mechanism",
                "type": "string"
              },
              "password": {
                "order": 3,
                "secret": true,
                "title": "Password",
                "type": "string"
              },
              "username": {
                "order": 2,
                "title": "Username",
                "type": "string"
              }
            },
            "required": [
              "auth_type",
              "mechanism",
              "password",
              "username"
            ],
            "title": "SASL (User & Password)"
          }
        ],
        "order": 1,
        "title": "Credentials",
        "type": "object"
      },
      "message_format": {
        "description": "Format for materialized messages. Avro format requires a schema registry configuration. Messages in JSON format do not use a schema registry.",
        "enum": [
          "Avro",
          "JSON"
        ],
        "order": 3,
        "title": "Message Format",
        "type": "string"
      },
      "schema_registry": {
        "description": "Connection details for interacting with a schema registry. This is necessary for materializing messages with Avro encoding.",
        "order": 5,
        "properties": {
          "endpoint": {
            "description": "Schema registry API endpoint. For example: https://registry-id.us-east-2.aws.confluent.cloud",
            "order": 0,
            "title": "Schema Registry Endpoint",
            "type": "string"
          },
          "password": {
            "description": "Schema registry password to use for authentication. If you are using Confluent Cloud, this will be the 'Secret' from your schema registry API key.",
            "order": 2,
            "secret": true,
            "title": "Schema Registry Password",
            "type": "string"
          },
          "username": {
            "description": "Schema registry username to use for authentication. If you are using Confluent Cloud, this will be the 'Key' from your schema registry API key.",
            "order": 1,
            "title": "Schema Registry Username",
            "type": "string"
          }
        },
        "required": [
          "endpoint",
          "password",
          "username"
        ],
        "title": "Schema Registry",
        "type": "object"
      },
      "tls": {
        "default": "system_certificates",
        "description": "Controls how should TLS certificates be found or used.",
        "enum": [
          "system_certificates"
        ],
        "order": 2,
        "title": "TLS Settings",
        "type": "string"
      },
      "topic_partitions": {
        "default": 6,
        "description": "The number of partitions to create new topics with.",
        "order": 6,
        "title": "Topic Partitions",
        "type": "integer"
      },
      "topic_replication_factor": {
        "default": 3,
        "description": "The replication factor to create new topics with.",
        "order": 7,
        "title": "Topic Replication Factor",
        "type": "integer"
      }
    },
    "required": [
      "bootstrap_servers",
      "credentials"
    ],
    "title": "Kafka Materialization Configuration",
    "type": "object"
  },
  "documentationUrl": "https://go.estuary.dev/materialize-kafka",
  "protocol": 3032023,
  "resourceConfigSchema": {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "properties": {
      "topic": {
        "description": "Name of the Kafka topic to materialize to.",
        "title": "Topic",
        "type": "string",
        "x-collection-name": true
      }
    },
    "required": [
      "topic"
    ],
    "title": "Kafka Resource Configuration",
    "type": "object"
  }
}
